File: warrior/src/main/kotlin/com/dumbdogdiner/warrior/user/UserCache.java
package com.dumbdogdiner.warrior.user;

import com.dumbdogdiner.warrior.Warrior;
import com.dumbdogdiner.warrior.WithWarriorPlugin;
import com.dumbdogdiner.warrior.api.user.WarriorUser;
import com.dumbdogdiner.warrior.api.user.WarriorUserCache;
import com.google.common.base.Preconditions;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.TreeMap;
import java.util.UUID;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public class UserCache implements WithWarriorPlugin, WarriorUserCache<User> {

    private final TreeMap<UUID, User> users = new TreeMap<>();

    public boolean contains(UUID uuid) {
        return users.containsKey(uuid);
    }

    public User addUser(@NotNull UUID userId) {
        Player p = Preconditions.checkNotNull(Bukkit.getPlayer(userId), "Player cannot be null!");

        if(!users.containsKey(userId)) {
            User user = new User(userId);

            Warrior.connection.insertUser(user);
            users.put(userId, user);

            return user;
        } else {
            String msg = String.format("Attempted to register WarriorUser %s{%2s} twice!", p.getName(), p.getUniqueId());
            Warrior.pluginLogger.warn(msg);
            return get(userId);
        }
    }

    public void remove(@NotNull User user) {
        remove(user.getUserId());
    }

    public void remove(@NotNull UUID userId) {
        if(users.containsKey(userId)) {
            User user = users.get(userId);
            user.saveData();

            users.remove(userId);
        } else {
            String msg = String.format("Failed to remove WarriorUser %s from cache.", userId);
            Warrior.pluginLogger.warn(msg);
        }
    }

    public User get(UUID userId) {
        return Preconditions.checkNotNull(users.get(userId), "WarriorUser is null!");
    }

    public List<User> getList() {
        return new ArrayList<>(users.values());
    }

    public List<User> getListOf(Predicate<User> predicate) {
        return users.values().stream().filter(predicate).collect(Collectors.toList());
    }
}

File: warrior/src/main/kotlin/com/dumbdogdiner/warrior/user/UserCache.kt
package com.dumbdogdiner.warrior.user

import com.dumbdogdiner.warrior.WarriorPlugin
import com.dumbdogdiner.warrior.WithWarriorPlugin
import com.dumbdogdiner.warrior.api.user.IUserCache
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.event.Listener
import java.util.*
import java.util.function.Predicate
import java.util.stream.Collectors
import kotlin.collections.ArrayList
import kotlin.collections.HashMap

class UserCache: IUserCache<User>, WithWarriorPlugin {

    private var userListener: Listener? = null
    var userMap: HashMap<UUID, User> = HashMap()

    override fun getListener(): Listener? {
        return userListener
    }

    override fun setListener(listener: Listener) {
        if(userListener == null) {
            this.userListener = listener

            this.unregister()
            Bukkit.getPluginManager().registerEvents(this.userListener!!, WarriorPlugin.instance)
        } else {
            getLogger().warn("Player Listener already registered... Skipping")
        }
    }

    override fun contains(uuid: UUID): Boolean {
        return userMap[uuid] != null
    }

    override fun addUser(userId: UUID): User? {
        val p: Player = Bukkit.getPlayer(userId)!!

        if(!p.isOnline) {
            getLogger().warn("User ${p.name}(${p.uniqueId}) is not online!")
            return null
        }

        return if(!this.contains(userId)) {
            val user = User(userId)
            // Need to add SQL stuff here later
            userMap[userId] = user
            user
        } else {
            getLogger().warn("Attempted to register WarriorUser ${p.name}(${p.uniqueId}) twice!")
            get(userId)
        }
    }

    override fun remove(user: User) {
        this.remove(user.uniqueId)
    }

    override fun remove(userId: UUID) {
        if(this.contains(userId)) {
            userMap.remove(userId)
        }
    }

    override fun get(userId: UUID): User? {
        return userMap[userId]
    }

    override fun getList(): MutableList<User> {
        return ArrayList(userMap.values)
    }

    override fun getListOf(predicate: Predicate<User>?): MutableList<User> {
        return this.list.stream().filter(predicate).collect(Collectors.toList())
    }

}